openvdb_ax/openvdb_ax/codegen/StandardFunctions.cc

```diff
@@ -2716,6 +2716,323 @@ inline FunctionGroup::UniquePtr axsort(const FunctionOptions& op)
 ///////////////////////////////////////////////////////////////////////////
 ///////////////////////////////////////////////////////////////////////////
 
+// Colour
+
+inline FunctionGroup::UniquePtr axhsvtorgb(const FunctionOptions& op)
+{
+    auto generate =
+        [op](const std::vector<llvm::Value*>& args,
+           llvm::IRBuilder<>& B) -> llvm::Value*
+    {
```

Comment:
Okay.... This is a really, really, strong argument to either support functions already in AX, or to create a plain text scripts that drive this building.

Reading code that assembles an LLVM parse tree is very hard.

The history of vex was first to make .vrw files that were plain text that could be parsed to generate the correct LLVM sequences, written at exactly this sort of level.  Here for example is the code to find the maximum of an array:

```
max@F[F (0) VEX_OPVANILLA {
    arraylength@I[F $it $1
    set@II $ii i(0)
    set@FF $0 f(-1e8)
    bwhile
$label1
    lt@III $ic $ii $it
    while@LI $label2 $ic
        getcomp@F[FI $ft $1 $ii
        max@FFF $0 $0 $ft
        add@II $ii i(1)
    jump@L $label1
$label2
}
```
This is then translated into the LLVM sequence points.

But, we then added the ability to just define these intrinsics as functions in the language itself, which made life infinitely better....

This rant prompted because I thought I'd just look at what function you implemented, but I can't have any idea from this...

Please prefix with a block comment with the equation you intend this to do, either written in AX or in LLVM....

Reply:
Yeah I'm completely aware of this - it's on the roadmap but isn't a small amount of work. After working so much with IR it's become a bit too easy to fall into this trap. Additionally, because we're lacking this higher level support that you're referring to, it's **essential** these are written in IR over C++ bindings for optimisation purposes. 

To combat this issue, AX is able to take both an IR implementation **and** a C/C++ binding. If you look through `StandardFunctions.cc` you'll see a number of cases where both a C++ function pointer and an IR implementation exist. We're starting to move away from this because direct IR (JIT) implementations **always** make more sense if they exist, but this "solves" the issues you're having because you can look at the C binding to infer what the IR is doing. Further more, for bindings to other libraries, we can just test that the result from the IR implementation matches. 

In summary, the C++ implementation is used in the unit tests to validate the IR and you can use that for a more readable implementation. See the diff for the unit tests: https://github.com/AcademySoftwareFoundation/openvdb/pull/1109/files

hsvtorgb: https://github.com/AcademySoftwareFoundation/openvdb/pull/1109/files#diff-a4df75455d40759f98e9279a80fffdc008a579eab7bb32af72ebccbe06e37e3bR663
rgbtohsv: https://github.com/AcademySoftwareFoundation/openvdb/pull/1109/files#diff-a4df75455d40759f98e9279a80fffdc008a579eab7bb32af72ebccbe06e37e3bR1181

And to confirm, I am working on frontend function support which is a re-requisite to what we're discussing here.